src/raftstore/store/snap/reader.rs

```diff
@@ -0,0 +1,274 @@
+            return Ok(0);
+        }
+        let inner = &mut self.inner;
+        while self.cf_index < inner.snapshot_meta.get_cf_files().len() {
+            if stale_for_generate(inner.key, self.snap_stale_notifier.as_ref()) {
+                let err = io::Error::new(io::ErrorKind::Other, "stale snapshot".to_owned());
+                return Err(err);
+            }
+
+            let cf = &inner.snapshot_meta.get_cf_files()[self.cf_index];
+            if cf.get_size() > 0 {
+                if self.cf_file.is_none() {
+                    let (dir, for_send, key) = (&inner.dir, inner.for_send, inner.key);
+                    let cf_file_path = get_cf_file_path(dir, for_send, key, cf.get_cf());
+                    self.cf_file = Some(File::open(&cf_file_path)?);
+                }
+                match self.cf_file.as_mut().unwrap().read(buf) {
+                    Ok(0) => {}
+                    Ok(n) => return Ok(n),
+                    e => return e,
+                }
+            }
+            self.cf_index += 1;
+            self.cf_file = None;
+        }
+        Ok(0)
+    }
+}
+
+pub(super) struct SnapshotLoader(Inner);
+
+impl SnapshotLoader {
+    pub(super) fn new(
+        dir: String,
+        for_send: bool,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+    ) -> Self {
+        let inner = Inner::new(dir, for_send, key, snapshot_meta);
+        SnapshotLoader(inner)
+    }
+
+    pub(super) fn load(mut self) -> Result<SnapshotMeta> {
+        for cf in self.0.snapshot_meta.get_cf_files() {
+            if cf.get_size() == 0 {
+                continue;
+            }
+            let (dir, for_send, key) = (&self.0.dir, self.0.for_send, self.0.key);
+            let cf_file_path = get_cf_file_path(dir, for_send, key, cf.get_cf());
+
+            let checksum = calc_crc32(&cf_file_path)?;
+            if checksum != cf.get_checksum() {
+                return Err(snapshot_checksum_corrupt(cf.get_checksum(), checksum));
+            }
+
+            let size = get_file_size(&cf_file_path)?;
+            if size != cf.get_size() {
+                return Err(snapshot_size_corrupt(cf.get_size(), size));
+            }
+        }
+        Ok(mem::replace(&mut self.0.snapshot_meta, SnapshotMeta::new()))
+    }
+}
+
+pub(super) struct SnapshotApplyer {
+    inner: Inner,
+    snap_stale_notifier: Arc<SnapStaleNotifier>,
+    applied: Arc<AtomicBool>,
+}
+
+impl SnapshotApplyer {
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+        snap_stale_notifier: Arc<SnapStaleNotifier>,
+        applied: Arc<AtomicBool>,
+    ) -> Self {
+        let inner = Inner::new(dir, false, key, snapshot_meta);
+        SnapshotApplyer {
+            inner,
+            snap_stale_notifier,
+            applied,
+        }
+    }
+
+    pub(super) fn apply(&self, options: ApplyOptions) -> Result<()> {
+        for cf in self.inner.snapshot_meta.get_cf_files() {
```

Comment:
I think should check the checksum here before applying.

Reply:
fixed.